okio-behavior/src/main/java/okio/test/BehaviorSink.java

```diff
@@ -0,0 +1,62 @@
+package okio.test;
+
+import java.io.IOException;
+import okio.Buffer;
+import okio.ForwardingSink;
+import okio.Sink;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+/** A wrapper which applies a {@link Behavior} to a {@link Sink}. */
+public final class BehaviorSink extends ForwardingSink {
+  private final Behavior behavior = new Behavior();
+
+  public BehaviorSink(Sink delegate) {
+    super(delegate);
+  }
+
+  public Behavior behavior() {
+    return behavior;
+  }
+
+  @Override public void write(Buffer source, long byteCount) throws IOException {
+    ByteRate rate = behavior.byteRate();
+
+    long written = 0L;
+    while (written < byteCount) {
+      long rateByteCount = rate.byteCount();
+      long rateSleepMs = rate.period(MILLISECONDS);
+
+      long periodByteCount = Math.min(rateByteCount, byteCount - written);
+
+      long readStartNanos = System.nanoTime();
+      super.write(source, periodByteCount);
+      long readTookMs = NANOSECONDS.toMillis(System.nanoTime() - readStartNanos);
+
+      written += periodByteCount;
+
+      long sleepMs = Math.max(0L, rateSleepMs - readTookMs);
+      if (sleepMs > 0L) {
+        try {
+          Thread.sleep(rateSleepMs);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          return;
+        }
+      }
+    }
+  }
+
+  @Override public void flush() throws IOException {
+    behavior.beforeFlush();
+    super.flush();
+    behavior.afterFlush();
```

Comment:
I added documentation to the fact that exceptions will result in later methods not being called. The main point of them is to throw exceptions, and I don't want to deal with the exception in `finally` case unless you feel very strongly.